File: java/src/leetcode/easy/FirstBadVersion.java
package leetcode.easy;

/**
 * Question: https://leetcode.com/problems/first-bad-version/description/
 */
public class FirstBadVersion {

    private int badVersion = 0;

    public FirstBadVersion(int badVersion) {
        this.badVersion = badVersion;
    }

    public int firstBadVersion(int n) {
        return firstBadVersion(1, n);
    }

    private int firstBadVersion(int start, int end) {
        if (start > end) return start;
        int middle = (int) (((long) start + end) / 2);
        if (isBadVersion(middle)) {
            return firstBadVersion(start, middle - 1);
        } else {
            return firstBadVersion(middle + 1, end);
        }
    }

    private boolean isBadVersion(int n) {
        return n == this.badVersion;
    }
}

File: ruby/easy/first_bad_version.rb
# The is_bad_version API is already defined for you.
# @param {Integer} version
# @return {boolean} whether the version is bad
# def is_bad_version(version):

# @param {Integer} n
# @return {Integer}
def first_bad_version(n)
    return binary_search(1, n)
end

def binary_search(start, stop)
    return start if start > stop
    middle = (start + stop)/2
    if(is_bad_version(middle))
        return binary_search(start, middle-1)
    else
        return binary_search(middle+1, stop)
    end
end
